File: advent2017/extras/cxx/src/days/day24.cpp
#include <chrono>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>
#include <openssl/sha.h>

namespace {
constexpr const char* EXPECTED_SHA = "48a139f917d7dac161171c28f578d923b212c10108c92bbe05a971f6d8b4fb05";
constexpr const char* EXPECTED_PART1 = "1656";
constexpr const char* EXPECTED_PART2 = "1642";

struct Comp {
    int a, b;
};
struct Ret {
    int s, l, ls;
};

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    std::vector<std::string> cands = {
        "advent2017/Day24/d24_input.txt",
        "Day24/d24_input.txt",
        "../Day24/d24_input.txt",
        "../../Day24/d24_input.txt",
    };
    for (const auto& c : cands) {
        std::ifstream f(c);
        if (f.good()) return c;
    }
    throw std::runtime_error("input not found");
}

std::string sha256_file(const std::string& path) {
    std::ifstream f(path, std::ios::binary);
    if (!f) throw std::runtime_error("read fail");
    std::vector<unsigned char> b((std::istreambuf_iterator<char>(f)), {});
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(b.data(), b.size(), out);
    static const char* HEX = "0123456789abcdef";
    std::string s;
    s.reserve(64);
    for (unsigned char v : out) {
        s.push_back(HEX[v >> 4]);
        s.push_back(HEX[v & 15]);
    }
    return s;
}

void parse(const std::string& path, std::vector<Comp>& comps, std::unordered_map<int, std::vector<int>>& by) {
    std::ifstream f(path);
    if (!f) throw std::runtime_error("open fail");
    std::string line;
    while (std::getline(f, line)) {
        if (line.empty()) continue;
        auto slash = line.find('/');
        int a = std::stoi(line.substr(0, slash));
        int b = std::stoi(line.substr(slash + 1));
        int i = static_cast<int>(comps.size());
        comps.push_back(Comp{a, b});
        by[a].push_back(i);
        if (b != a) by[b].push_back(i);
    }
}

Ret dfs(int port, const std::vector<Comp>& comps, const std::unordered_map<int, std::vector<int>>& by,
        std::vector<char>& used) {
    Ret best{0, 0, 0};
    auto it = by.find(port);
    if (it == by.end()) return best;
    for (int i : it->second) {
        if (used[i]) continue;
        const auto& c = comps[i];
        int nxt = (c.a == port) ? c.b : c.a;
        int seg = c.a + c.b;
        used[i] = 1;
        Ret ch = dfs(nxt, comps, by, used);
        used[i] = 0;

        if (seg + ch.s > best.s) best.s = seg + ch.s;
        int cl = 1 + ch.l;
        int cls = seg + ch.ls;
        if (cl > best.l || (cl == best.l && cls > best.ls)) {
            best.l = cl;
            best.ls = cls;
        }
    }
    return best;
}
}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part")
            part = std::stoi(argv[++i]);
        else if (a == "--input")
            input = argv[++i];
        else
            throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");
    input = resolve_input(input);
    if (sha256_file(input) != EXPECTED_SHA) throw std::runtime_error("checksum mismatch");

    std::vector<Comp> comps;
    std::unordered_map<int, std::vector<int>> by;
    parse(input, comps, by);
    std::vector<char> used(comps.size(), 0);

    auto t0 = std::chrono::steady_clock::now();
    Ret r = dfs(0, comps, by, used);
    std::string ans = std::to_string(part == 1 ? r.s : r.ls);
    std::string expected = part == 1 ? EXPECTED_PART1 : EXPECTED_PART2;
    if (ans != expected) throw std::runtime_error("answer mismatch");
    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=24 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day24.rs
// Architecture Notes:
// - This file is heavily commented for long-term maintainability and reconstruction.
// - Pipeline shape is parse -> model -> compute -> emit.
// - Performance-sensitive sections document data-layout and concurrency tradeoffs.

use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "48a139f917d7dac161171c28f578d923b212c10108c92bbe05a971f6d8b4fb05";
const EXPECTED_PART1: &str = "1656";
const EXPECTED_PART2: &str = "1642";

#[derive(Clone, Copy)]
struct Comp {
    a: i32,
    b: i32,
}

#[derive(Clone, Copy)]
struct Ret {
    s: i32,
    l: i32,
    ls: i32,
}

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() {
        return provided.to_string();
    }
    let cands = [
        "advent2017/Day24/d24_input.txt",
        "Day24/d24_input.txt",
        "../Day24/d24_input.txt",
        "../../Day24/d24_input.txt",
    ];
    for c in cands {
        if Path::new(c).exists() {
            return c.to_string();
        }
    }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new();
    h.update(&b);
    hex::encode(h.finalize())
}

fn parse(text: &str) -> (Vec<Comp>, HashMap<i32, Vec<usize>>) {
    let mut comps = Vec::new();
    let mut by: HashMap<i32, Vec<usize>> = HashMap::new();
    for ln in text.lines() {
        let s = ln.trim();
        if s.is_empty() {
            continue;
        }
        let p = s.split('/').collect::<Vec<&str>>();
        let a: i32 = p[0].parse().expect("int");
        let b: i32 = p[1].parse().expect("int");
        let i = comps.len();
        comps.push(Comp { a, b });
        by.entry(a).or_default().push(i);
        if b != a {
            by.entry(b).or_default().push(i);
        }
    }
    (comps, by)
}

fn solve(comps: &[Comp], by: &HashMap<i32, Vec<usize>>) -> (i32, i32) {
    let mut used = vec![false; comps.len()];
    fn dfs(port: i32, comps: &[Comp], by: &HashMap<i32, Vec<usize>>, used: &mut [bool]) -> Ret {
        let mut best = Ret { s: 0, l: 0, ls: 0 };
        if let Some(cands) = by.get(&port) {
            for &i in cands {
                if used[i] {
                    continue;
                }
                let c = comps[i];
                let nxt = if c.a == port { c.b } else { c.a };
                let seg = c.a + c.b;
                used[i] = true;
                let ch = dfs(nxt, comps, by, used);
                used[i] = false;

                if seg + ch.s > best.s {
                    best.s = seg + ch.s;
                }
                let cl = 1 + ch.l;
                let cls = seg + ch.ls;
                if cl > best.l || (cl == best.l && cls > best.ls) {
                    best.l = cl;
                    best.ls = cls;
                }
            }
        }
        best
    }
    let r = dfs(0, comps, by, &mut used);
    (r.s, r.ls)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => {
                part = args[i + 1].parse().unwrap();
                i += 2;
            }
            "--input" => {
                input = args[i + 1].clone();
                i += 2;
            }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 {
        panic!("--part 1|2 required");
    }
    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA {
        panic!("checksum mismatch");
    }
    let (comps, by) = parse(&fs::read_to_string(&in_path).expect("read_to_string"));

    let t0 = Instant::now();
    let (p1, p2) = solve(&comps, &by);
    let ans = if part == 1 {
        p1.to_string()
    } else {
        p2.to_string()
    };
    let expected = if part == 1 {
        EXPECTED_PART1
    } else {
        EXPECTED_PART2
    };
    if ans != expected {
        panic!("answer mismatch");
    }
    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=24 part={} runtime_ms={:.3}", part, ms);
}
